### PR TITLE
[protobuf+tools] refactor protobuf and make proto/grpc tooling consistent

### PR DESCRIPTION
**Pull request by truthbk**
### What does this PR do?

This PR aims to make protobuf easier to work with across different packages, and hopefully allow for reusing of messages whenever/wherever needed. The PR also adds:
- consistency to the tooling, we need to generate the protobuf with very specific versions due to a pin we have on our grpc code (that comes from a third-party dependency with `etcd`).
- a task to generate the protbuf messages and grpc code.

### Motivation

This PR is split off from the work in #8193 and provides a sane PR to review the protobuf refactor. Working on protobuf could be a little painful prior to this work. 

### Additional Notes

Anything else we should know when reviewing?

### Describe how to test your changes

Write here in detail how you have tested your changes
and instructions on how this should be tested in QA.

### CODE REVIEW

File: internal/tools/tools.go [hunk 1]
```diff
@@ -15,17 +15,23 @@ import (
...
 	_ "github.com/go-enry/go-license-detector/v4/cmd/license-detector"
+	_ "github.com/golang/mock/mockgen"
+	_ "github.com/golang/protobuf/protoc-gen-go"
 	_ "github.com/golangci/golangci-lint/cmd/golangci-lint"
 	_ "github.com/gordonklaus/ineffassign"
 	_ "github.com/goware/modvendor"
+	_ "github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway"
 	_ "github.com/mgechev/revive"
+	_ "github.com/stormcat24/protodep"
+	_ "golang.org/x/lint/golint"
```
Comment (mx-psi) [inline internal/tools/tools.go:26]:
```suggestion
```

File: internal/tools/go.mod [hunk 2]
```diff
@@ -6,14 +6,40 @@ require (
...
+	github.com/sirupsen/logrus v1.8.1 // indirect
+	github.com/sourcegraph/go-diff v0.6.1 // indirect
+	github.com/spf13/cobra v1.1.3 // indirect
+	github.com/spf13/viper v1.7.1 // indirect
+	github.com/stormcat24/protodep v0.0.0-20210106175019-c95189bc461b
+	github.com/timakin/bodyclose v0.0.0-20200424151742-cb6215831a94 // indirect
+	github.com/ultraware/funlen v0.0.3 // indirect
+	golang.org/x/crypto v0.0.0-20210220033148-5ea612d1eb83 // indirect
+	golang.org/x/exp v0.0.0-20200331195152-e8c3332aa8e5 // indirect
+	golang.org/x/lint v0.0.0-20200302205851-738671d3881b
```
Comment (mx-psi) [inline internal/tools/go.mod:36]: We don't use golint anymore
```suggestion
```

File: internal/tools/go.mod [hunk 3]
```diff
@@ -6,14 +6,40 @@ require (
 	github.com/client9/misspell v0.3.4
 	github.com/frapposelli/wwhrd v0.4.0
 	github.com/fzipp/gocyclo v0.3.1
+	github.com/go-critic/go-critic v0.5.6 // indirect
 	github.com/go-enry/go-license-detector/v4 v4.2.0
-	github.com/golangci/golangci-lint v1.40.1
+	github.com/gofrs/flock v0.8.0 // indirect
+	github.com/golang/mock v1.5.0
+	github.com/golang/protobuf v1.3.2
+	github.com/golangci/golangci-lint v1.22.2
```
Comment (mx-psi) [inline internal/tools/go.mod:14]: Please revert this change, we want to use the latest version

File: internal/tools/go.mod [hunk 4]
```diff
@@ -6,14 +6,40 @@ require (
 	github.com/client9/misspell v0.3.4
 	github.com/frapposelli/wwhrd v0.4.0
 	github.com/fzipp/gocyclo v0.3.1
+	github.com/go-critic/go-critic v0.5.6 // indirect
```
Comment (mx-psi) [inline internal/tools/go.mod:9]: There are a lot of indirect dependencies that we probably don't need to add. Can you review? If we need to add any, put them in a separate `require` block with a comment that explains the need

File: internal/tools/go.mod [hunk 5]
```diff
@@ -6,14 +6,40 @@ require (
...
+	github.com/ultraware/funlen v0.0.3 // indirect
+	golang.org/x/crypto v0.0.0-20210220033148-5ea612d1eb83 // indirect
+	golang.org/x/exp v0.0.0-20200331195152-e8c3332aa8e5 // indirect
+	golang.org/x/lint v0.0.0-20200302205851-738671d3881b
 	golang.org/x/mobile v0.0.0-20201217150744-e6ae53a27f4f
-	golang.org/x/perf v0.0.0-20200918155509-d949658356f9
+	golang.org/x/perf v0.0.0-20210220033136-40a54f11e909
 	golang.org/x/tools v0.1.2-0.20210512205948-8287d5da45e4
+	google.golang.org/genproto v0.0.0-20200128133413-58ce757ed39b // indirect
+	google.golang.org/grpc v1.28.0-pre
```
Comment (mx-psi) [inline internal/tools/go.mod:41]: Why do we need to add this?

File: tasks/test.py [hunk 6]
```diff
@@ -33,26 +34,55 @@ def ensure_bytes(s):
...
+]
+
 
 @task
 def install_tools(ctx):
     """Install all Go tools for testing."""
-    with ctx.cd("internal/tools"):
-        for tool in TOOL_LIST:
-            ctx.run("go install {}".format(tool))
+    with environ({'GO111MODULE': 'on'}):
```
Comment (mx-psi) [inline tasks/test.py:74]: Why do we need this?

File: tasks/test.py [hunk 7]
```diff
@@ -33,26 +34,55 @@ def ensure_bytes(s):
...
     'github.com/gordonklaus/ineffassign',
     'github.com/goware/modvendor',
     'github.com/mgechev/revive',
+    'github.com/stormcat24/protodep',
+    'golang.org/x/lint/golint',
     'gotest.tools/gotestsum',
     'honnef.co/go/tools/cmd/staticcheck',
 ]
 
+TOOL_LIST_GET = [
```
Comment (mx-psi) [inline tasks/test.py:63]: Why do we need this? I would like to avoid this: it goes against Go best practices, it does not let us use Dependabot to bump our tooling and it breaks the purpose of go mod tidy checks. If there is some sort of conflict with the versions on `internal/tools/go.mod` we should create a new `internal/tools/proto` module with these versions instead.

**Review by mx-psi — CHANGES_REQUESTED** [review]
Moving over some of the comments from #8193

File: internal/tools/go.mod (hunk 4 above)
Comment (sgnn7) [reply internal/tools/go.mod:9]: Maybe it just needs a `go mod tidy`? I could be wrong but I also think different versions of golang add these differently too so it would be good to do that with the exact version we use now too (v1.15.11).

File: pkg/proto/datadog/api/README.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+```
+This should drop all required binaries in your `$GOPATH/bin`
+
+Remember to make sure `GOPATH/bin` is in your `PATH`, also make
+sure no other versions of those binaries you may have installed
+elsewhere take precedence (`which` is your friend).
+
+### Code Generation
+
+Chdir yourself into this directory (`cmd/agent/api/pb`), and run
```
Comment (sgnn7) [inline pkg/proto/datadog/api/README.md:24]: Wrong path

File: pkg/proto/datadog/api/README.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+- protobuf - protoc-gen-go: generates the golang protobuf definitions.
+- grpc-gateway - protoc-gen-grpc-gateway: generates the gRPC-REST gateway  
+- grpc-gateway - protoc-gen-swagger (optional)
+
+### Install
+
+Run the following:
+```
+go install github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway github.com/grpc-ecosystem/grpc-gateway/protoc-gen-swagger github.com/golang/protobuf/protoc-gen-go
```
Comment (sgnn7) [inline pkg/proto/datadog/api/README.md:14]: I'm guessing we want them to run the tools install instead?

File: pkg/proto/datadog/api/README.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+```
+This should drop all required binaries in your `$GOPATH/bin`
+
+Remember to make sure `GOPATH/bin` is in your `PATH`, also make
+sure no other versions of those binaries you may have installed
+elsewhere take precedence (`which` is your friend).
```
Comment (sgnn7) [inline pkg/proto/datadog/api/README.md:20]: Is this valid still given that we use modules?

File: pkg/proto/datadog/api/v1/api.proto [hunk 11]
```diff
@@ -1,13 +1,19 @@
 syntax = "proto3";
 
-package pb;
+package datadog.api.v1;
 
+import "datadog/model/model.proto";
 import "google/api/annotations.proto";
 
+
+// option go_package = "github.com/DataDog/datadog-agent/pkg/proto/pbgo"; // golang
```
Comment (sgnn7) [inline pkg/proto/datadog/api/v1/api.proto:9]: Needs cleanup?

File: pkg/proto/datadog/api/v1/api.proto [hunk 12]
```diff
@@ -1,13 +1,19 @@
...
+option go_package = "pkg/proto/pbgo"; // golang
+
+
```
Comment (sgnn7) [inline pkg/proto/datadog/api/v1/api.proto:12]: nit: no need for extra linefeed

File: pkg/proto/datadog/model/model.proto [hunk 13]
```diff
@@ -0,0 +1,100 @@
+syntax = "proto3";
```
Comment (sgnn7) [inline pkg/proto/datadog/model/model.proto:1]: `api/v1/api.proto` is versioned and depends on this model so it would make sense that this proto is versioned too otherwise they are likely to fall out of sync.

File: pkg/proto/protodep.lock [hunk 14]
```diff
@@ -0,0 +1,31 @@
+proto_outdir = "./protodep"
+
+[[dependencies]]
+  target = "github.com/mwitkow/go-proto-validators"
+  revision = "32a686adf8b5194d3ea07d632d49b6fb344af678"
+  branch = "master"
+  path = "github.com/mwitkow/go-proto-validators"
```
Comment (sgnn7) [inline pkg/proto/protodep.lock:7]: I wonder if maybe we should add versions in the path for these? Might be also valuable to add a comment about what exact version matches the `revision`s for easier maintenance.

File: pkg/proto/utils/tagger.go [hunk 15]
```diff
@@ -0,0 +1,87 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+package utils
```
Comment (sgnn7) [inline pkg/proto/utils/tagger.go:6]: nit: Mixing a util into a `pkg/proto` package seems unnecessary and potentially bad for maintenance - why not move this to `pkg/tagger` (or `pkg/tagger/util/`)?

File: tasks/go.py [hunk 16]
```diff
@@ -438,6 +440,57 @@ def is_excluded(pkg):
...
+    with ctx.cd(repo_root):
+        # protobuf defs
+        print("generating protobuf code from: {}".format(proto_root))
+
+        files = []
+        for path in Path(os.path.join(proto_root, "datadog")).rglob('*.proto'):
+            files.append(path.as_posix())
+
+        ctx.run(
+            "protoc -I{include_path} --go_out=plugins=grpc:{out_path} {targets}".format(
```
Comment (sgnn7) [inline tasks/go.py:469]: Not something to be fixed now but we really need to allow `ctx run` to take in an array of strings instead of just a string.

File: tasks/test.py [hunk 17]
```diff
@@ -33,26 +34,55 @@ def ensure_bytes(s):
     return s
 
 
+@contextmanager
+def environ(env):
+    original_environ = os.environ.copy()
+    os.environ.update(env)
+    yield
```
Comment (sgnn7) [inline tasks/test.py:41]: I am a bit puzzled by this `yield` here - what is it supposed to do?

File: tasks/test.py [hunk 18]
```diff
@@ -33,26 +34,55 @@ def ensure_bytes(s):
...
-        for tool in TOOL_LIST:
-            ctx.run("go install {}".format(tool))
+    with environ({'GO111MODULE': 'on'}):
+        with ctx.cd("internal/tools"):
+            for tool in TOOL_LIST:
+                ctx.run("go install {}".format(tool))
+
+            for tool in TOOL_LIST_GET:
+                pkg = tool.split('@')[0]
+                ctx.run("go get {}".format(tool))
```
Comment (sgnn7) [inline tasks/test.py:81]: `go get --mod=readonly <tool>` so you don't need the tidy below.

PS: I also could be wrong, but I think you just aren't supposed to use `go install` nowadays with go modules but maybe Pablo can correct me. I think [there's fixes for 1.16](https://github.com/golang/go/issues/40276) but I haven't followed along the changes recently.

File: tasks/test.py [hunk 19]
```diff
@@ -33,26 +34,55 @@ def ensure_bytes(s):
...
+                ctx.run("go install {}".format(pkg))
```
Comment (sgnn7) [inline tasks/test.py:82]: the double `go install` for sure seems wrong

**Review by sgnn7 — CHANGES_REQUESTED** [review]
@truthbk Left some comments - definitely seems like the right track for better organization of this stuff!

File: tasks/test.py (hunk 17 above)
Comment (truthbk) [reply tasks/test.py:41]: `environ` is to be used as a context manager; the part above the yield is the setup, the part below it os the cleanup.

File: tasks/test.py (hunk 18 above)
Comment (truthbk) [reply tasks/test.py:81]: Oh yes, way better.  Indeed this can be done with go 1.16 simply as `go install some/pkg/foo@vX.Y.Z` we can hopefully clean this up surely after.

File: pkg/proto/datadog/api/README.md (hunk 8 above)
Comment (truthbk) [reply pkg/proto/datadog/api/README.md:24]: I need to cleanup the entire README, thanks for pointing this out.

File: pkg/proto/datadog/api/README.md (hunk 10 above)
Comment (truthbk) [reply pkg/proto/datadog/api/README.md:20]: probably not; will address with the readme cleanup.

File: tasks/test.py (hunk 19 above)
Comment (truthbk) [reply tasks/test.py:82]: Yes, this is collateral from a rebase, will cleanup.

File: tasks/test.py (hunk 6 above)
Comment (truthbk) [reply tasks/test.py:74]: Well, because the default value for that env var varies across go versions, it can also be left behind if the user messes with the environment, so I figured this was the right thing to do. I wish `invoke` had a nicer way to address this.

File: tasks/test.py (hunk 7 above)
Comment (truthbk) [reply tasks/test.py:63]: We need this because we need those pins. Indeed I was unable to get this to work properly with `internal/tools`, I'm fine with the `internal/tools/proto` approach.

File: internal/tools/proto/go.mod [hunk 20]
```diff
@@ -0,0 +1,9 @@
+module github.com/DataDog/datadog-agent/internal/tools/proto
+
+go 1.16
+
+require (
+	github.com/golang/mock v1.5.0 // indirect
+	github.com/golang/protobuf v1.3.2 // indirect
+	github.com/grpc-ecosystem/grpc-gateway v1.12.2 // indirect
```
Comment (mx-psi) [inline internal/tools/proto/go.mod:8]: I think none of these should have the `// indirect` comment since they are explicitly imported on the `tools.go` file

```suggestion
	github.com/golang/mock v1.5.0
	github.com/golang/protobuf v1.3.2
	github.com/grpc-ecosystem/grpc-gateway v1.12.2
```

File: tasks/go.py [hunk 21]
```diff
@@ -438,6 +440,57 @@ def is_excluded(pkg):
...
+                include_path=proto_root, out_path=repo_root, targets=' '.join(files),
+            )
+        )
+        # mockgen
+        mockgen_in = os.path.join(proto_root, "pbgo")
+        mockgen_out = os.path.join(proto_root, "pbgo", "mocks")
+        try:
+            os.mkdir(mockgen_out)
+        except FileExistsError:
+            pass
```
Comment (mx-psi) [inline tasks/go.py:485]: nit for easier debugging
```suggestion
            print("{} folder already exists".format(mockgen_out))
```

File: internal/tools/tools.go [hunk 22]
```diff
@@ -15,10 +15,12 @@ import (
 	_ "github.com/frapposelli/wwhrd"
 	_ "github.com/fzipp/gocyclo"
 	_ "github.com/go-enry/go-license-detector/v4/cmd/license-detector"
+	_ "github.com/golang/mock/mockgen"
```
Comment (mx-psi) [inline internal/tools/tools.go:18]: This is added twice, since you install it from the `proto` Go module I think we can remove it from here
```suggestion
```

(and run `go mod tidy` afterwards)

File: internal/tools/tools.go [hunk 23]
```diff
@@ -15,10 +15,12 @@ import (
...
 	_ "github.com/golangci/golangci-lint/cmd/golangci-lint"
 	_ "github.com/gordonklaus/ineffassign"
 	_ "github.com/goware/modvendor"
 	_ "github.com/mgechev/revive"
+	_ "github.com/stormcat24/protodep"
```
Comment (mx-psi) [inline internal/tools/tools.go:23]: I am fine leaving this here but I am curious if there is a reason for it to be here instead of on the `proto` module

File: internal/tools/go.mod [hunk 24]
```diff
@@ -7,12 +7,14 @@ require (
...
 	github.com/go-enry/go-license-detector/v4 v4.2.0
+	github.com/golang/mock v1.5.0
 	github.com/golangci/golangci-lint v1.40.1
 	github.com/gordonklaus/ineffassign v0.0.0-20210225214923-2e10b2664254
 	github.com/goware/modvendor v0.3.0
 	github.com/mgechev/revive v1.0.6
+	github.com/stormcat24/protodep v0.0.0-20210106175019-c95189bc461b
 	golang.org/x/mobile v0.0.0-20201217150744-e6ae53a27f4f
-	golang.org/x/perf v0.0.0-20200918155509-d949658356f9
+	golang.org/x/perf v0.0.0-20210220033136-40a54f11e909
```
Comment (mx-psi) [inline internal/tools/go.mod:17]: For future reference, why do we bump this?

File: internal/tools/go.mod (hunk 4 above)
Comment (mx-psi) [reply internal/tools/go.mod]: > Maybe it just needs a go mod tidy?

We have a `go mod tidy` check on the CI that passed so I think that would not have solved it (IIRC things pinned with `// indirect` will always be kept by `go mod tidy`).

> PS: Top of the file here interestingly mentions 1.16 as the go version for this module file

This *should* not have any undesirable effects but I will check it and open a separate PR to address this if needed

File: pkg/proto/protodep.toml [hunk 25]
```diff
@@ -0,0 +1,23 @@
+proto_outdir = "./protodep"
```
Comment (mx-psi) [inline pkg/proto/protodep.toml:1]: Is this file is generated by `protoc`? This looks like a Gopkg file, I suppose we can't use a `go.mod` file instead?

File: internal/tools/tools.go (hunk 23 above)
Comment (truthbk) [reply internal/tools/tools.go:23]: So we get the latest `protodep`, otherwise the pins in `internal/tools/proto` create issues.

File: internal/tools/go.mod (hunk 24 above)
Comment (truthbk) [reply internal/tools/go.mod:17]: I think I just `go mod tidy` after removing the stuff, I didn't bump that specifically.

File: pkg/proto/protodep.toml (hunk 25 above)
Comment (truthbk) [reply pkg/proto/protodep.toml:1]: Nope, the file is manually created and maintained and is specific to `protodep`.

File: internal/tools/proto/go.mod (hunk 20 above)
Comment (truthbk) [reply internal/tools/proto/go.mod:8]: Oh, I added those with `go get` to set the pins, otherwise it defaulted to other versions. But yeah, we don't need the `indirect` directives.

**Review by mx-psi — APPROVED** [review]
Tooling changes LGTM

File: pkg/proto/datadog/api/README.md [hunk 26]
```diff
@@ -0,0 +1,56 @@
+## gRPC: Protobuf and Gateway code generation 
+
+To generate the code for the API you have defined in your `.proto`
+files we will need three different grpc-related packages: 
+
+- protobuf - protoc-gen-go: generates the golang protobuf definitions.
+- grpc-gateway - protoc-gen-grpc-gateway: generates the gRPC-REST gateway
```
Comment (wiyu) [inline pkg/proto/datadog/api/README.md:7]: Would this be required for message only protos with no services?

File: tasks/go.py (hunk 16 above)
Comment (wiyu) [reply tasks/go.py:469]: I assume this is using whatever local version of `protoc` that is installed. Does the version of `protoc` installed matter or effect proto generation?

**Review by juliogreff — APPROVED** [review]

File: tasks/go.py (hunk 16 above)
Comment (truthbk) [reply tasks/go.py:469]: See: 
- https://github.com/DataDog/datadog-agent/tree/jaime/protobuf_refactor/internal/tools/proto
- https://github.com/DataDog/datadog-agent/blob/c5ce993c5ec133f57584302d2d35009477826bfa/tasks/test.py#L63-L67

File: pkg/proto/datadog/api/README.md (hunk 26 above)
Comment (truthbk) [reply pkg/proto/datadog/api/README.md:7]: If there's nothing to generate, the tool won't do anything for protobuf messages not related with any grpc service. This use-case is already addressed correctly by this PR.

File: pkg/proto/utils/tagger.go (hunk 15 above)
Comment (truthbk) [reply pkg/proto/utils/tagger.go:6]: We could try that, but I think having these here avoids the likelihood of dependency cycles; something which also motivated this PR.

File: pkg/proto/protodep.lock (hunk 14 above)
Comment (truthbk) [reply pkg/proto/protodep.lock:7]: This file is auto-generated by protodep based of `pkg/proto/protodep.toml` where the revisions are explicit.

**Review by wiyu — APPROVED** [review]

File: pkg/proto/datadog/api/README.md [hunk 27]
```diff
@@ -0,0 +1,56 @@
+## gRPC: Protobuf and Gateway code generation
```
Comment (sgnn7) [inline pkg/proto/datadog/api/README.md:1]: I can't seem to add a comment about it since it's just a rename but [`pkg/proto/datadog/README.md`](https://github.com/DataDog/datadog-agent/blob/c5ce993c5ec133f57584302d2d35009477826bfa/pkg/proto/datadog/README.md) has some outdated info on it now that it moved.

File: tasks/test.py (hunk 17 above)
Comment (sgnn7) [reply tasks/test.py:41]: I would probably guess that we want a try/except/finally pattern around the yield then to ensure we clean up things but it's not a blocker.

**Review by sgnn7 — APPROVED** [review]
@truthbk Just a few small nit comments but LGTM as-is too.
